simplex/docs/reconfiguration.md

```diff
@@ -0,0 +1,471 @@
+The start time of epoch $E + 1$ is defined to be the build time of B<sup>E</sup><sub>end</sub>.
+
+The purpose of ICM epochs is to aggregate validator changes across all L1s into coarse grained units of reference.
+
+Unlike ICM epochs, A Simplex epoch doesn't change unless the validator set of the L1 changes,
+as its role is to convey the validator set that is used to validate the quorum certificate of the block.
+
+Further distinguishing themselves from ICM epochs, Simplex epoch numbers aren't successive, but are defined to be the sequence numbers of the blocks that seal them.
+More concretely, let $B_k$ be the last block of epoch $e$.
+Then, the next epoch is numbered $k$ and not $e+1$.
+
+We denote such a block $B_k$ with sequence number $k$ which is part of epoch $e$ as the $sealing~block$ of epoch $e$.
+The validator set for epoch $k$ is derived from the P-chain height of the sealing block $B_k$.
+
+In order to support ICM epochs, the Simplex protocol facilitates the encoding of both ICM epochs and Simplex epochs.
+Whenever a block is built, the block builder encodes the ICM epoch information (`EpochStartTime`, `epoch_number`, `PChainEpochHeight`) in the same manner as the proposerVM encodes it,
+and also encodes the Simplex epoch information, as well as some additional auxiliary information that may be used by other protocols, as explained below:
+
+```proto
+message SimplexEpochInfo {
+   uint64 p_chain_reference_height = 1;
+   uint64 epoch_number = 2;
+   uint64 next_p_chain_reference_height = 3;
+   uint64 prev_vm_block_seq = 4; // The sequence of the previous VM block
+   uint64 sealing_block_seq = 5; // The sequence number of the sealing block
+   BlockValidationDescriptor block_validation_descriptor = 6; // Describes how to validate the blocks of the next epoch
+   NextEpochApprovals next_epoch_approvals = 7; // The epoch change approvals of the next epoch by at least n-f nodes.
+}
+```
+
+- The validator set of the epoch numbered `epoch_number` is derived from `p_chain_reference_height`.
+
+- The `next_p_chain_reference_height` is the P-chain height of the next epoch, otherwise it is set to `0`.
+
+- The `prev_vm_block_seq` is the sequence number of the previous VM block, and it is used to efficiently find the last VM block upon startup.
+
+- The `sealing_block_seq` is the sequence number of the sealing block of the epoch, and it is set to `0` before the sealing block is created.
+  It is used to determine when the Simplex instance can transition to the next epoch. It is safe to transition to the next epoch once the sealing block is finalized.
+
+- The `block_validation_descriptor` describes how to validate the blocks of the next epoch.
+  It is encoded in blocks starting from the block that has `next_p_chain_reference_height > 0` until and including the sealing block.
+  It is used to verify the quorum certificates of blocks that are built in the next epoch.
```

Comment:
Can we make it so that we only include this in the sealing block? It seems like a bunch of duplicate information (which we will need to verify the correctness of in every block)... Additionally, if we ever wanted to support threshold simplex, we might not even know this until after the DKG has finished.

Additionally, if this is only provided in the sealing block, then `sealing_block_seq` doesn't seem to be needed, as it is seemingly only used as a bool `sealing_block_seq > 0`.

Reply:
We could but then it would be very heavy/complex to validate `next_epoch_approvals`.

It's there since `next_p_chain_reference_height` is defined because you validate the `next_epoch_approvals` mutation by looking at the `block_validation_descriptor` of the previous block and using it. 

I guess we could add some kind of caching mechanism so we don't need to always re-compute `block_validation_descriptor`. 

> Additionally, if we ever wanted to support threshold simplex, we might not even know this until after the DKG has finished.

Yes that's a good point.